[Improvement-11857][Spark] Remove the spark version of spark task

rickchengx opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler. Please review https://dolphinscheduler.apache.org/en-us/community/development/pull-request.html before opening a pull request.-->


## Purpose of the pull request

close: #11857 

Currently, the spark version is misleading.
<img width="577" alt="截屏2022-09-08 15 24 34" src="https://user-images.githubusercontent.com/38122586/189061020-9198f272-9239-4f32-a671-530c5b66a05e.png">

This is not the spark version that DS currently supports. E.g., DS can also run the spark task of `SPARK-3.x.x`.

The spark version selected by the user only determines the environment variables used and the final command executed as below:

<img width="640" alt="image" src="https://user-images.githubusercontent.com/38122586/189061737-4674445c-072f-4705-a891-f96f9ea70d61.png">

And also there is a bug that `spark-sql` can only be executed by `SPARK2`, see https://github.com/apache/dolphinscheduler/pull/11721

So why not just remove the spark version, like other task
And use `{SPARK_HOME}/bin/spark_submit` and `{SPARK_HOME}/bin/spark-sql`

## Brief change log

 use `{SPARK_HOME}/bin/spark_submit` and `{SPARK_HOME}/bin/spark-sql`

## Verify this pull request

<!--*(Please pick either of the following options)*-->

manually tested

<img width="578" alt="image" src="https://user-images.githubusercontent.com/38122586/189084759-ab2f3fb1-d7b6-4ad8-a916-28e16586c7a2.png">


<img width="1021" alt="截屏2022-09-08 16 51 41" src="https://user-images.githubusercontent.com/38122586/189083321-837ca066-7f29-43a6-a5a9-abe723feee2f.png">

rickchengx committed:
[Improvement-11857][Spark] Remove the spark version of spark task

rickchengx commented:
Related Doc will be modified after this PR is reviewed

SbloodyS commented:
> Related Doc will be modified after this PR is reviewed

Please add the docs in this PR. We suggest that the documents involved in the modification should be submitted together.

rickchengx commented:
> > Related Doc will be modified after this PR is reviewed
> 
> Please add the docs in this PR. We suggest that the documents involved in the modification should be submitted together.

Sure, but there are many documents that need to be modified, maybe we can first review whether this PR is appropriate

SbloodyS commented:
> > > Related Doc will be modified after this PR is reviewed
> > 
> > 
> > Please add the docs in this PR. We suggest that the documents involved in the modification should be submitted together.
> 
> Sure, but there are many documents that need to be modified, maybe we can first review whether this PR is appropriate

That sound good to me.

fuchanghai commented:
If there is only one version of spark on a machine, there is no problem keeping only one. Problems can arise if a machine has two versions. Of course, I don't know about spark, I just analyze this problem from a business perspective

fuchanghai commented:
@SbloodyS may be this PR is conflict with #11720 ,According to the actual application scenario, it seems that the current PR should be supported. if ds want to support two different versions ,i support another

rickchengx commented:
Hi, @fuchanghai, Thanks for the reply. I agree with your point. 
1. I think that the current `spark version` on ui is kind of **misleading**. Users may be concerned that DS does not support `spark 3.x.x`. I have seen people asking whether DS supports spark3 , even if there is a description for this problem in the DS [documentation](https://dolphinscheduler.apache.org/en-us/docs/dev/user_doc/guide/installation/kubernetes.html) (but users will not find this description easily) as below:

<img width="811" alt="截屏2022-09-09 09 55 13" src="https://user-images.githubusercontent.com/38122586/189256318-8c9cbd85-a6db-4662-981b-57df74774136.png">


2. Even if DS needs to support multiple sparks, I don't think setting `SPARK_HOME1` and `SPARK_HOME2` is an elegant way. What if the user needs to support 3 or more sparks?

3. I think that using `SPARK_HOME` is an easy and efficient way without misunderstandings and some potential problems.

SbloodyS commented:
> @SbloodyS may be this PR is conflict with #11720 ,According to the actual application scenario, it seems that the current PR should be supported. if ds want to support two different versions ,i support another

I think it's better to use current environment management to manage different version of spark. This is the meaning of environmental management.

rickchengx commented:
> > @SbloodyS may be this PR is conflict with #11720 ,According to the actual application scenario, it seems that the current PR should be supported. if ds want to support two different versions ,i support another
> 
> I think it's better to use current environment management to manage different version of spark. This is the meaning of environmental management.


@SbloodyS Thanks for your suggestion. I totally agree with you.

So DS uses `SPARK_HOME/bin/spark-submit` to build the command, and user can set the `SPARK_HOME` he wants in the environment.

<img width="587" alt="截屏2022-09-09 10 13 33" src="https://user-images.githubusercontent.com/38122586/189257831-118719b3-f360-44fb-88e3-44dc10a96e2a.png">

SbloodyS commented:
> > > @SbloodyS may be this PR is conflict with #11720 ,According to the actual application scenario, it seems that the current PR should be supported. if ds want to support two different versions ,i support another
> > 
> > 
> > I think it's better to use current environment management to manage different version of spark. This is the meaning of environmental management.
> 
> @SbloodyS Thanks for your suggestion. I totally agree with you.
> 
> So DS uses `SPARK_HOME/bin/spark-submit` to build the command, and user can set the `SPARK_HOME` he wants in the environment.
> 
> <img alt="截屏2022-09-09 10 13 33" width="587" src="https://user-images.githubusercontent.com/38122586/189257831-118719b3-f360-44fb-88e3-44dc10a96e2a.png">

Yes. That's what I mean.

rickchengx committed:
update the docs related to spark version

rickchengx commented:
> Yes. That's what I mean.

Hi, @SbloodyS , @fuchanghai 

Since #11721 is merged, I've rebased my PR and also modified the related docs.

So now this PR uses `SPARK_HOME/bin/spark-submit` or `SPARK_HOME/bin/spark-sql` to build the command, and user can set the `SPARK_HOME` he wants in the environment as below.

<img width="570" alt="截屏2022-09-13 14 40 34" src="https://user-images.githubusercontent.com/38122586/189828791-345bf3b0-fe57-4d0a-92f1-3100d9928ff0.png">


So the misleading `spark version` is removed, and different tasks can still use different `spark versions` through the environment.

zhongjiajie commented on docs/docs/en/guide/task/spark.md:
```diff
@@ -15,34 +15,33 @@ Spark task type for executing Spark application. When executing the Spark task,
-| Predecessor task | Selecting a predecessor task for the current task, will set the selected predecessor task as upstream of the current task. |
+|       **Parameter**        |                                                                                                        **Description**                                                                                                        |
+|----------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Node Name                  | Set the name of the task. Node names within a workflow definition are unique.                                                                                                                                                 |
+| Run flag                   | Indicates whether the node can be scheduled normally. If it is not necessary to execute, you can turn on the prohibiting execution switch.                                                                                    |
+| Description                | Describes the function of this node.                                                                                                                                                                                          |
+| Task priority              | When the number of worker threads is insufficient, they are executed in order from high to low according to the priority, and they are executed according to the first-in, first-out principle when the priority is the same. |
+| Worker group               | The task is assigned to the machines in the worker group for execution. If Default is selected, a worker machine will be randomly selected for execution.                                                                     |
+| Task group name            | The group in Resources, if not configured, it will not be used.                                                                                                                                                               |
+| Environment Name           | Configure the environment in which to run the script.                                                                                                                                                                         |
+| Number of failed retries   | The number of times the task is resubmitted after failure. It supports drop-down and manual filling.                                                                                                                          |
+| Failure Retry Interval     | The time interval for resubmitting the task if the task fails. It supports drop-down and manual filling.                                                                                                                      |
+| Timeout alarm              | Check Timeout Alarm and Timeout Failure. When the task exceeds the "timeout duration", an alarm email will be sent and the task execution will fail.                                                                          |
```
It is seems we already change the general parameter into separate parameter files, can you rebase to latest and take a look?

rickchengx replied:
Sure.

zhongjiajie reviewed (commented):
please rebase the latest code, our docs make some change
and can you also add some discription in both https://github.com/apache/dolphinscheduler/blob/4dca488cd50b4392d222167c01ae2a79fd295e77/dolphinscheduler-python/pydolphinscheduler/UPDATING.mdand https://github.com/apache/dolphinscheduler/blob/3aa9f2ea25ca42112141aad85140a72b0963e2c3/docs/docs/en/guide/upgrade/incompatible.md becaus this is incompatible change for  users

rickchengx commented:
> please rebase the latest code, our docs make some change and can you also add some discription in both https://github.com/apache/dolphinscheduler/blob/4dca488cd50b4392d222167c01ae2a79fd295e77/dolphinscheduler-python/pydolphinscheduler/UPDATING.mdand https://github.com/apache/dolphinscheduler/blob/3aa9f2ea25ca42112141aad85140a72b0963e2c3/docs/docs/en/guide/upgrade/incompatible.md becaus this is incompatible change for users

Hi, @zhongjiajie , thanks for your comment. I've rebased my PR and add some descriptions in `UPDATING.md` and `incompatible.md`

zhongjiajie reviewed (approved):
LGTM, thanks

zhongjiajie commented:
@caishunfeng should we release this PR in version 3.1.0?
